---
title: "Quality Control of sc samples"
output: 
  html_document
---

The first notebook will introduce you the quality control concepts that will be critical for detecting the "good" cells from the "bad" cells. QC is the most critical step in any single-cell workflow and you should ideally spend at least 70% of your time on quality checking the data.

We have a saying, "crap in crap out". Bioinformatic software will always give you an answer no matter the quality of the data. You need to be smart about what is fed into the software so you can be confident of the biological interpretability.

The first thing we will do is load the required packages:

```{r, cache=FALSE, warning = FALSE, message = FALSE}
knitr::opts_chunk$set(cache=FALSE, warning = FALSE, message = FALSE)
library(tidyverse)
library(SingleCellExperiment)
library(Matrix)
library(ggplot2)
library(scales)
library(tximport)
library(BUSpaRse)
library(Seurat)

library("biomaRt")

# Specify the sample directory that you would like analysed:
sample_files <- str_replace(Sys.glob("kallisto.dir/*"), "kallisto.dir/", "")


```


Next we will load the counts data into memory using the tximport package. In addition to this we will load the barcodes generated from salmon and will import a list of genes.

```{r}

for (i in sample_files){
  name <- paste("kallisto.dir/", i, "/bus/genecount/", sep="")
  res_mat <- read_count_output(name,
                             name = "genes", tcc = FALSE)
  # Get rid of version number on ensembl gene
  rownames(res_mat) <- gsub("\\.\\d+$","", rownames(res_mat))
  
  
  so <- CreateSeuratObject(res_mat, min.cells=10, min.features=50) %>% 
    NormalizeData(verbose=FALSE) %>% 
    ScaleData(verbose=FALSE) %>% 
    FindVariableFeatures(verbose=FALSE)
  
  counts <- Matrix::colSums(so@assays$RNA@counts)
  genes <- Matrix::colSums(so@assays$RNA@counts>0)
  
  metadata <- as.data.frame(so@meta.data)
  assign(paste("so", i, sep = "."), so)
  assign(paste("counts", i, sep = "."), counts)
  assign(paste("genes", i, sep = "."), genes)
  assign(paste("res_mat", i, sep = "."), res_mat)
  assign(paste("metadata", i, sep = "."), metadata)
  rm(so)
  rm(counts)
  rm(genes)
  rm(metadata)
}

```

```{r setup}
knitr::opts_chunk$set(echo = TRUE)

library(jsonlite)
library(tidyverse)

theme_Publication <- function(base_size=12, base_family="Arial") {
      library(grid)
      library(ggthemes)
      (theme_bw(base_size=base_size, base_family=base_family)
       + theme(plot.title = element_text(face = "bold",
                                         size = rel(1.2), hjust = 0.5),
               text = element_text(),
               panel.background = element_rect(colour = NA),
               plot.background = element_rect(colour = NA),
               panel.border = element_rect(colour = NA),
               axis.title = element_text(face = "bold",size = rel(1)),
               axis.title.y = element_text(angle=90,vjust =2),
               axis.title.x = element_text(),
               axis.text = element_text(), 
               axis.line = element_line(colour="black"),
               axis.ticks = element_line(),
               panel.grid.major = element_line(colour="#f0f0f0"),
               panel.grid.minor = element_blank(),
               legend.key = element_rect(colour = NA),
               legend.position = "bottom",
               legend.direction = "horizontal",
               legend.key.size= unit(0.2, "cm"),
               legend.margin = unit(0, "cm"),
               legend.title = element_text(face="italic"),
               plot.margin=unit(c(10,5,5,5),"mm"),
               strip.background=element_rect(colour="#f0f0f0",fill="#f0f0f0"),
               strip.text = element_text(face="bold")
          ))
      
}

scale_fill_Publication <- function(...){
      library(scales)
      discrete_scale("fill","Publication",manual_pal(values = c("#386cb0","#fdb462","#7fc97f","#ef3b2c","#662506","#a6cee3","#fb9a99","#984ea3","#ffff33")), ...)

}

scale_colour_Publication <- function(...){
      library(scales)
      discrete_scale("colour","Publication",manual_pal(values = c("#386cb0","#fdb462","#7fc97f","#ef3b2c","#662506","#a6cee3","#fb9a99","#984ea3","#ffff33")), ...)

}
```

# Creating metadata object with some QC metrics

Now that we have a counts matrix with the genes as row names and cells as columns, we can create our metadata with information about the different metrics to evaluate during quality control assessment.

We will create the metadata with only the cell IDs:

```{r}
for (i in sample_files){
  so <- get(gsub("SAMPLE_FILE",i , "so.SAMPLE_FILE"))
  counts <- get(gsub("SAMPLE_FILE",i , "counts.SAMPLE_FILE"))
  genes <- get(gsub("SAMPLE_FILE",i , "genes.SAMPLE_FILE"))
  metadata <- get(gsub("SAMPLE_FILE",i , "metadata.SAMPLE_FILE"))
  # Add number of UMIs for each gene per cell to metadata
  metadata$nUMI <- counts
  # Add number of genes detected per cell to metadata
  metadata$nGene <- genes
  # Add number of UMIs per gene for each cell to metadata
  metadata$log10GenesPerUMI <- log10(metadata$nGene) / log10(metadata$nUMI)
  # Add sample name associated with each cell to metadata 
  metadata$sample <- gsub("SAMPLE_FILE",i ,"SAMPLE_FILE")
  so@meta.data <- metadata
  assign(paste("metadata", i, sep = "."), metadata)
  assign(paste("so", i, sep = "."), so)
  rm(counts)
  rm(metadata)
}
```



# Generate additional QC metrics



```{r}

# useMart keeps not working every so often (issue https://github.com/grimbough/biomaRt/issues/31)
ensembl <- try(useMart("ensembl", dataset = "hsapiens_gene_ensembl"))
if(class(ensembl) == "try-error"){
  httr::set_config(httr::config(ssl_verifypeer = FALSE))
  ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
}

mapping <- getBM(attributes = c("ensembl_transcript_id_version","ensembl_gene_id_version","ensembl_gene_id", "hgnc_symbol"), mart = ensembl)
mito <- grep("^MT-", mapping$hgnc_symbol, value=TRUE)
mito <- mapping %>% filter(hgnc_symbol %in% mito)
mito <- mito$ensembl_gene_id
```


```{r}

for (i in sample_files){
  metadata <- get(gsub("SAMPLE_FILE",i , "metadata.SAMPLE_FILE"))
  counts <- get(gsub("SAMPLE_FILE",i , "counts.SAMPLE_FILE"))
  so <- get(gsub("SAMPLE_FILE",i , "so.SAMPLE_FILE"))
  metadata$mtUMI <- Matrix::colSums(so@assays$RNA@counts[which(rownames(so@assays$RNA@counts) %in% mito),], na.rm = T)
  
  # Calculate of mitoRatio per cell
  metadata$mitoRatio <- metadata$mtUMI/metadata$nUMI
  assign(paste("metadata", i, sep = "."), metadata)
  so@meta.data <- metadata
  assign(paste("so", i, sep = "."), so)
}


```



```{r hgnc}
# Map ensembl symbols to hgnc symbols using biomart

mapping2 <- distinct(mapping[,2:4]) # remove transcripts IDs

for (i in sample_files){
  metadata <- get(gsub("SAMPLE_FILE",i , "metadata.SAMPLE_FILE"))
  counts <- get(gsub("SAMPLE_FILE",i , "counts.SAMPLE_FILE"))
  so <- get(gsub("SAMPLE_FILE",i , "so.SAMPLE_FILE"))

  ensembl_genes_so <- data.frame(ensembl_gene_id = rownames(so))
  joined_ensembl_hgnc <- plyr::join(ensembl_genes_so, mapping2, by = "ensembl_gene_id", type = "left", match="first")
  # Matches first HGNC, if multiple HGNC map to one ensembl id. Keep the same order unlike merge function
  joined_ensembl_hgnc$hgnc_symbol <-  na_if(joined_ensembl_hgnc$hgnc_symbol, "")
  # Changes "" to NAs

  metadata$hgnc_symbols <- joined_ensembl_hgnc$hgnc_symbol
  assign(paste("metadata", i, sep = "."), metadata)
  so@meta.data <- metadata
  assign(paste("so", i, sep = "."), so)
}

```

# Filtering!!!

This is very important to get correct as any "bad cells" will ultimately end up ruining your downstream analysis.

## Initial filtering

Prior to assessing our metrics, we are going to perform a very minimal filtering of those cells with less than 100 UMIs to get rid of the cells that are clearly junk, containing less than 100 UMIs.

```{r}
for (i in sample_files){
  counts <- get(gsub("SAMPLE_FILE",i , "res_mat.SAMPLE_FILE"))
  metadata <- get(gsub("SAMPLE_FILE",i , "metadata.SAMPLE_FILE"))
# Keep cells with nUMI greater than 100
idx <- which(metadata$nUMI > 5)

# Extract the counts for those cells
counts_c <- counts[, idx]

# Extract the metadata for those cells
metadata_c <- metadata[idx,]
assign(paste("metadata_c", i, sep = "."), metadata)
}
```

# Saving metrics to single cell experiment

Before we assess our metrics we are going to save all of the work we have done thus far to a single cell experiment object, which is a standard object for single cell data in R (https://www.bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html).

```{r}
for (i in sample_files){
  so <- get(gsub("SAMPLE_FILE",i , "so.SAMPLE_FILE"))
  counts <- so@assays$RNA@counts
  metadata <- get(gsub("SAMPLE_FILE",i , "metadata.SAMPLE_FILE"))
# Save data to single cell experiment variable
se <- SingleCellExperiment(assays=list(counts=counts), 
                           colData = metadata)
dir.create("RDS_objects.dir", showWarnings = FALSE)
# Create .RData object to load at any time
saveRDS(so, gsub("SAMPLE_FILE",i ,"RDS_objects.dir/SAMPLE_FILE_unfiltered_SeuratObject.rds"))
saveRDS(se, gsub("SAMPLE_FILE",i ,"RDS_objects.dir/SAMPLE_FILE_raw_unfiltered_SingleCellExp.rds"))
assign(paste("se", i, sep = "."), se)
}
```

# Assessing the quality metrics

Now that we have generated the various metrics to assess, we can explore them with visualizations. We will create our metrics file from the metadata stored in the single cell experiments.

```{r}
for (i in sample_files){
  se <- get(gsub("SAMPLE_FILE",i , "se.SAMPLE_FILE"))
# Create a data frame containing the metrics for visualizations
metrics <- colData(se) %>%
  as.data.frame
assign(paste("metrics", i, sep = "."), metrics)
}
```

We will explore the following metrics through visualizations to decide on which cells are low quality and should be removed from the analysis:

* Cell counts
* UMI counts per cell
* Genes detected per cell
* UMIs vs. genes detected
* Mitochondrial counts ratio
* Novelty

# Visualize the number of cell counts per cell {.tabset .tabset-fade}

```{r, results='asis', echo = FALSE}
dir.create("QC_Figures.dir", showWarnings = FALSE)
for (i in sample_files){
  metrics <- get(gsub("SAMPLE_FILE",i , "metrics.SAMPLE_FILE"))
# Visualize the number of cell counts per cell

cat("## ",i,"\n")
plt <- metrics %>% 
  ggplot(aes(x=sample, fill=sample)) + 
  geom_bar() + 
  ggtitle("Number of Cells") + theme_Publication()
print(plt)

name<- paste0("QC_Figures.dir/NumberOfCells_", i, ".eps")
postscript(name)
print(plt)
dev.off()
cat('\n\n')
}
```

# UMI counts (transcripts) per cell {.tabset .tabset-fade}

The UMI counts per cell should generally be above 500, although usable, itâ€™s still low if between 500-1000 counts. If UMIs per cell is 500-1000 counts, then the cells probably should have been sequenced more deeply.

```{r, results='asis', echo = FALSE}
for (i in sample_files){
  metrics <- get(gsub("SAMPLE_FILE",i , "metrics.SAMPLE_FILE"))
# Visualize the number UMIs/transcripts per cell

cat("## ",i,"\n")
plt <- metrics %>% 
        ggplot(aes(color=sample, x=nUMI, fill= sample)) + 
        geom_density() + 
        scale_x_log10() + 
        ylab("log10 cell density") +
        geom_vline(xintercept = 50) + theme_Publication()
plot(plt)

name<- paste0("QC_Figures.dir/Log10CellDensity_", i, ".eps")
postscript(name)
plot(plt)
dev.off()
cat('\n\n')
}
```

# Genes detected per cell {.tabset .tabset-fade}

Seeing gene detection in the range of 500-5000 is normal for inDrop analysis. Similar expectations for gene detection as for UMI detection, although may be a bit lower than UMIs.

```{r, results='asis', echo = FALSE}
for (i in sample_files){
  metrics <- get(gsub("SAMPLE_FILE",i , "metrics.SAMPLE_FILE"))
# Visualize the distribution of genes detected per cell via histogram

cat("## ",i,"\n")
plt <- metrics %>% 
        ggplot(aes(color=sample, x=nGene, fill= sample)) + 
        geom_density() + 
        scale_x_log10() + 
        geom_vline(xintercept = 20) + theme_Publication()
print(plt)

name<- paste0("QC_Figures.dir/GenesDetectedDensity_", i, ".eps")
postscript(name)
plot(plt)
dev.off()
cat('\n\n')
}
```

# Ncells vs Ngenes {.tabset .tabset-fade}

```{r, results='asis', echo = FALSE}
for (i in sample_files){
  metrics <- get(gsub("SAMPLE_FILE",i , "metrics.SAMPLE_FILE"))
# Visualize the distribution of genes detected per cell via histogram

cat("## ",i,"\n")
# Visualize the distribution of genes detected per cell via boxplot
plt <- metrics %>% 
        ggplot(aes(x=sample, y=log10(nGene), fill=sample)) + 
        geom_boxplot() + 
        ggtitle("NCells vs NGenes") + theme_Publication()
print(plt)


name<- paste0("QC_Figures.dir/NcellsvsNgenes_", i, ".eps")
postscript(name)
plot(plt)
dev.off()
cat('\n\n')
}
```


# UMIs vs. genes detected {.tabset .tabset-fade}

Poor quality cells are likely to have low genes and UMIs per cell. Therefore, a poor sample is likely to have cells in the lower left of the graph. Good cells should exhibit both higher number of genes per cell and higher numbers of UMIs. We also expect similar lines with similar slopes for all samples.

```{r, results='asis', echo = FALSE}
for (i in sample_files){
  metrics <- get(gsub("SAMPLE_FILE",i , "metrics.SAMPLE_FILE"))
# Visualize the distribution of genes detected per cell via histogram

cat("## ",i,"\n")
# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
plt <- metrics %>% 
  ggplot(aes(x=nUMI, y=nGene, color=mitoRatio)) + 
  geom_point() + 
  stat_smooth(method=lm) +
  scale_x_log10() + 
  scale_y_log10() + 
  geom_vline(xintercept = 800) +
  facet_wrap(~sample) + theme_Publication()
print(plt)

name<- paste0("QC_Figures.dir/nUMIvsnGene_", i, ".eps")
postscript(name)
print(plt)
dev.off()
cat('\n\n')
}
```


# Mitochondrial counts ratio {.tabset .tabset-fade}

This metric can identify whether there is a large amount of mitochondrial contamination from dead or dying cells. Poor quality samples for mitochondrial counts would have larger peaks above the 0.1 mitochondrial ratio mark, unless it is expected based on sample type.

```{r, results='asis', echo = FALSE}
for (i in sample_files){
  metrics <- get(gsub("SAMPLE_FILE",i , "metrics.SAMPLE_FILE"))
# Visualize the distribution of genes detected per cell via histogram

cat("## ",i,"\n")
# Visualize the distribution of mitochondrial gene expression detected per cell
plt<- metrics %>% 
        ggplot(aes(color=sample, x=mitoRatio, fill=sample)) + 
        geom_density() + 
        scale_x_log10() + 
        geom_vline(xintercept = 0.3) + theme_Publication()
print(plt)


name<- paste0("QC_Figures.dir/MitoRatio_nGenevsnUMI", i, ".eps")
postscript(name)
print(plt)
dev.off()
cat('\n\n')
}
```

# Novelty {.tabset .tabset-fade}

We can see the samples where we sequenced each cell less have a higher overall novelty, that is because we have not started saturating the sequencing for any given gene for these samples. Outlier cells in these samples might be cells that have a less complex RNA species than other cells. Sometimes we can detect contamination with low complexity cell types like red blood cells via this metric. Generally, we expect the novelty score to be above 0.80.

```{r, results='asis', echo = FALSE}
for (i in sample_files){
  metrics <- get(gsub("SAMPLE_FILE",i , "metrics.SAMPLE_FILE"))
# Visualize the distribution of genes detected per cell via histogram

cat("## ",i,"\n")
# Visualize the overall novelty of the gene expression by visualizing the genes detected per UMI
plt <- metrics %>%
        ggplot(aes(x=log10GenesPerUMI, color = sample, fill=sample)) +
        geom_density() + theme_Publication()
print(plt)

name<- paste0("QC_Figures.dir/Novelty_", i, ".eps")
postscript(name)
print(plt)
dev.off()
cat('\n\n')
}
```
